spray-json/shared/src/main/scala/spray/json/JsonParser.scala

```diff
@@ -165,51 +170,148 @@ class JsonParser(input: ParserInput, settings: JsonParserSettings = JsonParserSe
+      } else
+        parseOneSlow(cursor + 1, charsRead + 1)
     }
 
-  private def `escaped`() = {
-    def hexValue(c: Char): Int =
-      if ('0' <= c && c <= '9') c - '0'
-      else if ('a' <= c && c <= 'f') c - 87
-      else if ('A' <= c && c <= 'F') c - 55
-      else fail("hex digit")
-    def unicode() = {
-      var value = hexValue(cursorChar)
-      advance()
-      value = (value << 4) + hexValue(cursorChar)
-      advance()
-      value = (value << 4) + hexValue(cursorChar)
-      advance()
-      value = (value << 4) + hexValue(cursorChar)
-      appendSB(value.toChar)
+    def escaped(cursor: Int, charsRead: Int): Int =
+      input.byteAt(cursor + 1) match {
+        case c @ ('"' | '/' | '\\') =>
+          charBuffer(charsRead) = c.toChar
+          2
+        case 'b' =>
+          charBuffer(charsRead) = '\b'
+          2
+        case 'f' =>
+          charBuffer(charsRead) = '\f'
+          2
+        case 'n' =>
+          charBuffer(charsRead) = '\n'
+          2
+        case 'r' =>
+          charBuffer(charsRead) = '\r'
+          2
+        case 't' =>
+          charBuffer(charsRead) = '\t'
+          2
+        case 'u' =>
+          def hexValue(c: Byte): Int =
+            if ('0' <= c && c <= '9') c - '0'
+            else if ('a' <= c && c <= 'f') c - 87
+            else if ('A' <= c && c <= 'F') c - 55
+            else fail("hex digit")
+
+          charBuffer(charsRead) =
+            ((hexValue(input.byteAt(cursor + 2)) << 12) |
+              (hexValue(input.byteAt(cursor + 3)) << 8) |
+              (hexValue(input.byteAt(cursor + 4)) << 4) |
+              hexValue(input.byteAt(cursor + 5))).toChar
+
+          6
+        case _ => fail("JSON escape sequence")
+      }
+    def unicodeChar(b: Char, cursor: Int, charsRead: Int): Int = {
+      if ((b & 0xE0) == 0xC0) { // two byte sequence
+        charBuffer(charsRead) = (((b & 0x1f) << 6) |
+          (input.byteAt(cursor + 1) & 0x3f)).toChar
+        2
+      } else if ((b & 0xF0) == 0xE0) { // 3-byte UTF-8 sequence
+        val codePoint =
+          ((b & 0x0f) << 12) |
+            ((input.byteAt(cursor + 1) & 0x3f) << 6) |
+            (input.byteAt(cursor + 2) & 0x3f)
+
+        if (codePoint < 0xffff) {
+          charBuffer(charsRead) = codePoint.toChar
+          3
+        } else {
+          charBuffer(charsRead) = ((0xd7C0 + (codePoint >> 10)).toChar)
+          charBuffer(charsRead + 1) = ((0xdc00 + (codePoint & 0x3ff)).toChar)
+          3
+        }
+      } else if ((b & 0xF8) == 0xF0) { // 4-byte UTF-8 sequence
+        val codePoint =
+          ((b & 0x07) << 18) |
+            ((input.byteAt(cursor + 1) & 0x3f) << 12) |
+            ((input.byteAt(cursor + 2) & 0x3f) << 6) |
+            (input.byteAt(cursor + 3) & 0x3f)
+
+        if (codePoint < 0xffff) {
+          charBuffer(charsRead) = codePoint.toChar
+          4
+        } else {
+          charBuffer(charsRead) = ((0xd7C0 + (codePoint >> 10)).toChar)
```

Comment:
I would add checking for broken surrogate pairs here, emoji so popular these days :smile: